src/Workspaces/Core/Portable/Serialization/SerializableSourceText.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Serialization
+{
+    internal sealed class SerializableSourceText
+    {
+        public ITemporaryTextStorageWithName? Storage { get; }
+        public SourceText? Text { get; }
+
+        public SerializableSourceText(ITemporaryTextStorageWithName storage)
+            : this(storage, text: null)
+        {
+        }
+
+        public SerializableSourceText(SourceText text)
+            : this(storage: null, text)
+        {
+        }
+
+        private SerializableSourceText(ITemporaryTextStorageWithName? storage, SourceText? text)
+        {
+            Debug.Assert(storage is not null || text is not null);
+
+            Storage = storage;
+            Text = text;
+        }
+
+        public ImmutableArray<byte> GetChecksum()
+        {
+            if (Storage is not null)
+                return Storage.GetChecksum();
+            else if (Text is not null)
+                return Text.GetChecksum();
+            else
+                return ImmutableArray<byte>.Empty;
+        }
+
+        public async ValueTask<SourceText> GetTextAsync(CancellationToken cancellationToken)
+        {
+            if (Text is not null)
+                return Text;
+
+            if (Storage is not { } textStorage)
```

Comment:
What does this achieve over just null checking Storage and using Storage directly below?

Or even just using `Storage!` given its an invariant of the class constructor.

Reply:
This was a type check in a previous revision. I'll simplify it.

Reply:
➡️ Done